### DISCUSSION TITLE
How to use cocotb from a larger build infrastructure?

### DISCUSSION THREAD

**Discussion by sbhutada** [cocotb/cocotb#3596] [Q&A] [answered]
I am trying to integrate pyuvm/cocotb inside larger build infrastructure. What is the best way to do this? I thought venv might work. Any suggestions? Thanks.
Installation steps:
```
cd $COMMON_ROOT/scripts/rtlutils
python -m venv uvmpy/venv
/usr/local/fbcode/platform010/bin/python3.8 -m venv uvmpy/venv
cd uvmpy/
source venv/bin/activate
git clone https://github.com/pyuvm/pyuvm
cd pyuvm
python -m pip install .
```

User usage steps:
```
setenv UVMPY $COMMON_ROOT/scripts/rtlutils/uvmpy
source $UVMPY/venv/bin/activate.csh
cp -rf $UVMPY/pyuvm/examples .
cd examples/TinyALU/
```
Update Makefile to add this to it:
include ${COMMON_ROOT}/scripts/rtlutils/pyuvm/cleanall.mk
```
make sim
Traceback (most recent call last):
  File ".../uvmpy/venv/bin/cocotb-config", line 5, in <module>
    from cocotb.config import main
  File ".../uvmpy/venv/lib/python3.8/site-packages/cocotb/__init__.py", line 43, in <module>
    import cocotb.handle
  File ".../uvmpy/venv/lib/python3.8/site-packages/cocotb/handle.py", line 39, in <module>
    from cocotb import simulator
ImportError: cannot import name 'simulator' from partially initialized module 'cocotb' (most likely due to a circular import) (.../uvmpy/venv/lib/python3.8/site-packages/cocotb/__init__.py)
Makefile:26: /Makefile.sim: No such file or directory
make: *** No rule to make target '/Makefile.sim'.  Stop.
```
Is there a way to use Makefiles without coco-tb-config?

**Comment by sbhutada** [comment]
I used 3.12 and it seems to work.

I just need to know how to avoid using cocotb-config - or how to integrate this in larger build infrastructure.

Thanks.

**Comment by ktbarrett** [comment, marked as answer]
This seems to just be an issue with your environment configuration.
```
python -m venv uvmpy/venv
/usr/local/fbcode/platform010/bin/python3.8 -m venv uvmpy/venv
```

Why are you making the venv twice with different Python versions? The venv is probably broken. Only do this once.

**Reply by sbhutada** [reply]
That command just creates venv under uvmpy/venv. Not twice. I could have also just done python -m venv .

**Comment by cmarqu** [comment]
In general, the documentation for your question is at https://docs.cocotb.org/en/stable/custom_flows.html

**Reply by sbhutada** [reply]
I tried it with python3.12 venv but libpython is failing since the python installation I am using doesn't have libpython.so.

.../python3.12/site-packages/cocotb/share/makefiles/Makefile.inc:159: *** 

find_libpython was not able to find a libpython in the current Python environment. Ensure
the Python development packages are installed. If they are installed and find_libpython
is not finding the path to libpython, file an upstream bug in find_libpython; then
manually override the LIBPYTHON_LOC make variable with the absolute path to libpython.so
(or python.dll on Windows).
.  Stop.

It looks like I need libpython.so. I don't have admin privileges on my network. So I am trying miniconda. But that doesn't have libpython.so either. What is the best way to get local python dev package?

**Comment by sbhutada** [comment]
Thanks. Fixed the issue with miniconda.
